DataArray: propagate index coordinates with non-array dimensions

benbovy opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] support DataArray constructor
- [x] support DataArray constructed from Dataset (variable access)
- [x] support update or `assign_coords`
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

This PR allows including coordinates in a DataArray that do not need to have dimensions strictly matching (a subset of) the array dimensions. It still requires such coordinates to be associated with an index that shares at least one dimension with the array dimensions.

This is particularly useful for CF cell boundaries coordinates, see #1475.

Here is an example [notebook](https://notebooksharing.space/view/91d4cf942b6466a02d1931047022294147475dd7500345fbb2545a068dd3490b#displayOptions=) showing this PR in action with an Xarray IntervalIndex that works with CF 2-dimensional bounds coordinates (#8005).

max-sixty commented on xarray/core/dataarray.py:
```diff
@@ -130,18 +130,30 @@
-def _check_coords_dims(shape, coords, dim):
-    sizes = dict(zip(dim, shape, strict=True))
+def _check_coords_dims(
+    shape: tuple[int, ...], coords: Coordinates, dims: tuple[Hashable, ...]
```
Only internal so low stress, but FYI we agreed to standardize on `dim` for param names (except for constructor)

benbovy commented:
Activated benchmarks on this PR. There is more logic now executed for the very common operation of dataarray access from dataset (e.g., `ds.foo`) involving `Indexes.get_all_dims()`, which may impact performance. I don't think we could really avoid it but hopefully the impact is small.

dcherian commented on xarray/core/coordinates.py:
```diff
@@ -964,22 +964,32 @@ def __getitem__(self, key: Hashable) -> T_DataArray:
-        original_indexes.update(indexes)
-        self._data._indexes = original_indexes
+        if set(dims) > set(self.dims):
+            for k, v in coords.items():
```
```suggestion
            for name, var in coords.items():
```
 
I found this very confusing since there's a dict comprehension that uses `k,v` below :)

dcherian commented on xarray/core/coordinates.py:
```diff
@@ -964,22 +964,32 @@ def __getitem__(self, key: Hashable) -> T_DataArray:
+        if set(dims) > set(self.dims):
+            for k, v in coords.items():
+                if any(d not in self.dims for d in v.dims):
+                    # allow any coordinate associated with an index that shares at least
+                    # one of dataarray's dimensions
+                    temp_indexes = Indexes(
+                        indexes, {k: v for k, v in coords.items() if k in indexes}
+                    )
+                    if k in indexes:
+                        index_dims = temp_indexes.get_all_dims(k)
+                        if any(d in self.dims for d in index_dims):
+                            continue
```
I don't understand this at all. Can we make it simpler? Or at least expand the comments. The repeated creation of `Indexes` seems like it can have a large impact.

dcherian commented on xarray/core/dataarray.py:
```diff
@@ -130,18 +130,30 @@
         if any(d not in dim for d in v.dims):
+            # allow any coordinate associated with an index that shares at least
+            # one of dataarray's dimensions
+            indexes = coords.xindexes
```
let's move this outside the loop.

dcherian commented on xarray/core/dataarray.py:
```diff
@@ -130,18 +130,30 @@
-def _check_coords_dims(shape, coords, dim):
+def _check_coords_dims(
+    shape: tuple[int, ...], coords: Coordinates, dim: tuple[Hashable, ...]
+):
```
I'd appreciate a docstring to understand what `dim` is.

dcherian commented on xarray/core/dataarray.py:
```diff
@@ -130,18 +130,30 @@
 
         for d, s in v.sizes.items():
-            if s != sizes[d]:
+            if d not in extra_index_dims and s != sizes[d]:
```
```suggestion
            if d not in extra_index_dims or s != sizes[d]:
```
?

benbovy replied:
This should be `and`.

`extra_index_dims` corresponds to all non-array dimensions of index coordinates to include in the dataarray (size[d] would return a KeyError).

I renamed it and added comment in [695fb86](https://github.com/pydata/xarray/pull/10116/commits/695fb8611c16d64a4f1b199e1324ecfc53592c5e).

dcherian commented on xarray/core/dataset.py:
```diff
@@ -1210,10 +1210,20 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
         for k in self._variables:
-            if k in self._coord_names and set(self._variables[k].dims) <= needed_dims:
-                coords[k] = self._variables[k]
+            if k in self._coord_names:
```
I'm surprised this is so much cleaner than the dataarray case (which has a nested loop). Is the approach here applicable to dataarray?

dcherian commented on xarray/tests/test_dataarray.py:
```diff
@@ -1634,6 +1658,31 @@ def test_assign_coords_no_default_index(self) -> None:
+
+        # cannot use `assert_identical()` test utility function here yet
+        # (indexes invariant check is still based on IndexVariable, which
+        # doesn't work with AnyIndex coordinate variables here)
```
Is the invariant test easy to update? It would be really nice to keep it up to date.

benbovy replied:
Done in [bc578d8](https://github.com/pydata/xarray/pull/10116/commits/bc578d846bc9a6f38644c49091ea39320144920a) (this was actually already update for the Dataset invariant checks).

dcherian commented:
I was really excited to see this PR! Thanks, Benoit.

Given that this is a fairly major data model change for DataArray, it'd be nice to get @shoyer's input too.

benbovy commented:
Thanks for the review @dcherian! I addressed most of your comments in [695fb86](https://github.com/pydata/xarray/pull/10116/commits/695fb8611c16d64a4f1b199e1324ecfc53592c5e), hopefully it is much clearer now.

I agree this PR needs further careful review from the Xarray core devs!

I didn't add any documentation for this change. There should be no user-facing change until we also provide custom indexes like IntervalIndex, so we should probably wait before advertising it.

@max-sixty I reverted using `dims` as argument name for the (renamed) `check_dataarray_coords` internal helper function to make it consistent with the other helper functions below that also use `dims`. Using `dims` in `check_dataarray_coords` makes senses to me since the value is always a tuple. No strong opinion, though, we can address this here or in a follow-up PR. (Could you point me to the discussion of using `dim` over `dims`, please?).

benbovy commented:
I also have to do some profiling, at least for the failing repr benchmark.

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -130,22 +130,59 @@
+
+    # check dimension names
+    for name, var in coords.items():
+        if name in skip_check_coord_name:
+            continue
+        elif name in indexes:
+            index_dims = indexes.get_all_dims(name)
+            if any(d in dims for d in index_dims):
+                # can safely skip checking index's non-array dimensions
+                # and index's other coordinates since those must be all
+                # included in the dataarray so the index is not corrupted
+                skip_check_coord_name.update(indexes.get_all_coords(name))
+                skip_check_dim_size.update(d for d in index_dims if d not in dims)
+                raise_error = False
+            else:
+                raise_error = True
+        else:
+            raise_error = any(d not in dims for d in var.dims)
+        if raise_error:
             raise ValueError(
-                f"coordinate {k} has dimensions {v.dims}, but these "
+                f"coordinate {name} has dimensions {var.dims}, but these "
                 "are not a subset of the DataArray "
-                f"dimensions {dim}"
+                f"dimensions {dims}"
             )
```
I'm having a hard time following the logic in this loop. Could you please separate it into two separate loops:
1. Update the sets of all coords/names to skip
2. Raise any necessary errors

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -130,22 +130,59 @@
+
+    for name, var in coords.items():
+        for d, s in var.sizes.items():
+            if d not in skip_check_dim_size and s != sizes[d]:
```
If I understand correctly, the only case in which a dimension is in `skip_check_dim_size` is when it is defined only on an index, in which case it won't appear in `sizes`.

In that case, I think we  could simplify this by checking `sizes` instead of creating the `skip_check_dim_size` set: 
```suggestion
            if d in sizes and s != sizes[d]:
```

dcherian commented:
>  It is really the worst-case scenario, with 100 data variables each having their own, single-index dimension coordinate.

Do we have a benchmark for O(300) vars with the only 4 dimensions? (this is the ERA5 dataset)

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1210,10 +1210,20 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
-            if k in self._coord_names and set(self._variables[k].dims) <= needed_dims:
-                coords[k] = self._variables[k]
+            if k in self._coord_names:
+                if (
+                    k not in coords
+                    and k in temp_indexes
+                    and set(temp_indexes.get_all_dims(k)) & needed_dims
+                ):
+                    # add all coordinates of each index that shares at least one dimension
+                    # with the dimensions of the extracted variable
+                    coords.update(temp_indexes.get_all_coords(k))
```
Coud this use a separate loop after the existing loop instead? e.g.,
```python
for k in self._indexes:
    if k in coords:
        coords.update(self.xindexes.get_all_coords(k))
```

Or if we allow indexes without a coordinate of the same name:
```python
for k in self._indexes:
    if set(self.xindexes.get_all_dims(k)) & needed_dims:
        coords.update(self.xindexes.get_all_coords(k))
```

Ideally, I would like the logic here to be just as simple as the words describing how it works, so a comment is not necessary!

dcherian commented:
Another thing to fix is that the Indexes section of the repr will need its own list of dimensions, since these are not all shared with DataArray.dims. This could be done in a future PR, but we should at least open an issue in that case.

benbovy commented:
That would useful to test (I can have a look), although I suspect that with 4 dimensions and no custom index (so maximum 4 default single-coordinate indexes) there won't be any significant impact.

shoyer commented:
Thanks @benbovy for putting this together. Custom indexes do seem like a great way to solve the coordinate propagation problem for cases like cell bounds.

One edge case that comes to mind are coordinates that share dimensions with an index dimension, but that are not associated with an index directly. For example, consider a `longitude` dimension with a `longitude_bounds` coordinate with dimensions `(longitude, longitude_bounds)` and another coordinate variable `flux` also with dimensions `(longitude, longitude_bounds)`. Would `flux` be included automatically in every case where `longitude` is preserved or not? In other words, does including coordinates from indexes happen before or after including all coordinates with dimensions associated with other variables/coordinates?

Another scenario I am concerned about is a 2D spatial index, e.g., along `latitude` and `longitude`. If the result needs only the `latitude` dimension, would the 2D spatial index and the `longitude` coordinate still be included? This would probably not be desirable.

A different way to implement the logic of this PR (in a more general way) would be to delegate deciding which coordinates to include to a custom method on the xarray `Index` class. This might be a good idea if we are not sure we can come up with a good general rule for all cases.

benbovy commented:
Thanks for your feedback @shoyer!

> Would flux be included automatically in every case where longitude is preserved or not? In other words, does including coordinates from indexes happen before or after including all coordinates with dimensions associated with other variables/coordinates?

In that case `flux` wouldn't be included but we could easily update the rules so that including non-indexed coordinates depends on a set of dimensions calculated after going over all indexed coordinates.

> Another scenario I am concerned about is a 2D spatial index, e.g., along latitude and longitude. If the result needs only the latitude dimension, would the 2D spatial index and the longitude coordinate still be included? This would probably not be desirable.

Yes indeed I can see how it would be confusing, in some cases desirable and in other cases not.

> A different way to implement the logic of this PR (in a more general way) would be to delegate deciding which coordinates to include to a custom method on the xarray Index class. This might be a good idea if we are not sure we can come up with a good general rule for all cases.

I didn't thought about it but I like this idea! On the plus side this would certainly prevent computing a set of dimensions for each index.

My main concern is that the `Index` API is already quite flexible (maybe one reason this feature hasn't been much adopted so far? but certainly not the main reason since documentation is still almost non-existent and built-in custom indexes are still missing...), and delegating even more to custom indexes might not help towards driving more adoption. That may be a minor concern, though.

I'm not sure yet how would look like such custom method on the `Index` class. Do you have any idea? Maybe something like this?

```python
class Index:

    def get_dataarray_coords(
        self,
        variables: dict[Hashable, Variable],
        dims: tuple[Hashable, ...],
    ) -> Coordinates:
        """Select or generate coordinates to be included in a DataArray.

        This method is called when a new DataArray is created either from scratch,
        by assigning new coordinates to an existing DataArray or by extracting a
        variable from a Dataset.

        All the returned coordinate variables (and their indexes) will be included
        in the new DataArray, whether or not their dimensions match (a subset of)
        the array dimensions.

        It is possible to return coordinates with another index instance (possibly of
        an other type) in case the new DataArray has a different (reduced) set of
        dimensions.

        Parameters
        ----------
        variables : dict
            All coordinate variables associated to this index.
        dims: tuple
            Dataarray's dimensions.

        Returns
        -------
        A new :py:class:`xarray.Coordinates` instance that will
        be included in the new DataArray (maybe alongside other
        coordinates).
      
        """
        # probably better to return something meaningful as the default implementation.
        raise NotImplementedError()
    
```


The `variables` argument is needed since The `Index` base class does not provide any built-in functionality or model that keeps track of coordinate variables (and their dimensions) associated with an index. The return type `Coordinates` allows maximum flexibility, e.g., return the index itself and all of its coordinates as we would do for an IntervalIndex with center and bounds coordinates, return a 1D PandasIndex from a 2D spatial index if the result is only the `latitude` dimension, etc.

Note that an index may already control which coordinates (and indexes) are propagated in a DataArray via other `Index` methods like `Index.isel`, `Index.sel`, etc. as shown by the IntervalIndex example linked in the top comment.

benbovy commented:
> probably better to return something meaningful as the default implementation.

It would make sense to return all coordinates and their index if their dimensions are all compatible with the dataarray dimensions, or otherwise raise an error (or drop the index return only the coordinates with compatible dimensions).

benbovy commented:
> A different way to implement the logic of this PR (in a more general way) would be to delegate deciding which coordinates to include to a custom method on the xarray Index class. This might be a good idea if we are not sure we can come up with a good general rule for all cases.

See #10137!
